# Pandas Tutorial

This is guide helps you quickly learn and use Python pandas for our course: **Python Interactive Dashboards with Plotly Dash**.

Pandas is the most popular Python package for data manipulation and analysis. Data analysts or data scientists often spend most of their time exploring and transforming data, making pandas an essential skill for data science.

In this tutorial, you’ll learn the following pandas techniques that will be used in the course:

- Installing and importing pandas
- Pandas data structures
- Importing CSV files
- Exploring the DataFrame
- Selecting and manipulating columns and rows
  - Selecting columns
  - Finding unique values in a column
  - Filtering rows
  - Group by

## Installing and importing pandas
If you've installed Python with Anaconda, you should already have pandas installed.

If you have Python but not pandas, use the command below within the terminal to install it.

```pip install pandas```



After installing the package, we need to import it for each new Python session.

In [1]:
import pandas as pd

## Pandas data structures
Pandas has two main data structures: `Series` and `DataFrame`.

`Series`: a 1-dimensional labeled array that can hold any data type such as integers, strings, floating points, Python objects.
It has row/axis labels as the index.

`DataFrame`: a 2-dimensional labeled data structure with columns of potentially different types.
It also has row labels as the index.

DataFrames can be considered as a collection of Series; it has a structure like a spreadsheet.

To work in pandas, we can either create DataFrames or read data from external sources. In the course, we will use CSV files as the data sources.

## Importing CSV files

The `read_csv` method is the standard way of importing CSV files into pandas DataFrames.

For example, here we read in `world_happiness.csv`, which gives us the happiness scores of different countries. 

Please make sure to save the CSV file in the same directory as the Python script. Otherwise, you will need to specify the path within the function.

In [2]:
happiness = pd.read_csv('world_happiness.csv')
happiness

country                     region  happiness_rank  \
0         Australia  Australia and New Zealand              10   
1         Australia  Australia and New Zealand               9   
2         Australia  Australia and New Zealand              10   
3         Australia  Australia and New Zealand              10   
4         Australia  Australia and New Zealand              11   
..              ...                        ...             ...   
847  United Kingdom             Western Europe              23   
848  United Kingdom             Western Europe              19   
849  United Kingdom             Western Europe              11   
850  United Kingdom             Western Europe              15   
851  United Kingdom             Western Europe              12   

     happiness_score  year  
0             7.2840  2015  
1             7.3130  2016  
2             7.2840  2017  
3             7.2720  2018  
4             7.2280  2019  
..               ...   ...  
847           6.7250  2016  
848           6.7140  2017  
849           7.1900  2018  
850           7.0540  2019  
851           7.1645  2020  

[852 rows x 5 columns]

## Exploring the DataFrame

We can also use some DataFrame methods to explore information about the dataset.

### info

The `info` method shows a helpful summary of our DataFrames. It is one of the first methods you should apply after loading in the dataset.

In [3]:
happiness.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852 entries, 0 to 851
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country          852 non-null    object 
 1   region           852 non-null    object 
 2   happiness_rank   852 non-null    int64  
 3   happiness_score  852 non-null    float64
 4   year             852 non-null    int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 33.4+ KB


### head

The `head` method is also used in every data analysis project. With its help, we can see exactly how the first five rows look like within the DataFrame.

In [4]:
happiness.head()

country                     region  happiness_rank  happiness_score  year
0  Australia  Australia and New Zealand              10            7.284  2015
1  Australia  Australia and New Zealand               9            7.313  2016
2  Australia  Australia and New Zealand              10            7.284  2017
3  Australia  Australia and New Zealand              10            7.272  2018
4  Australia  Australia and New Zealand              11            7.228  2019

### shape
To get the number of rows and columns of the DataFrame, we can use `shape`.

In [5]:
happiness.shape

(852, 5)

Based on the outputs above, we can see that `happiness` is a DataFrame with 852 rows and five columns. The column names are:
- `country`
- `region`
- `happiness_rank`
- `happiness_score`
- `year`

## Selecting and manipulating columns and rows

### Selecting columns
The columns of the DataFrames represent variables/features in the dataset. 

The most basic way of selecting columns is by calling their name within squared brackets ```[]```.

For example, we can print out the columns `country` and `region` separately below.

In [6]:
happiness['country']

0           Australia
1           Australia
2           Australia
3           Australia
4           Australia
            ...      
847    United Kingdom
848    United Kingdom
849    United Kingdom
850    United Kingdom
851    United Kingdom
Name: country, Length: 852, dtype: object

In [7]:
happiness['region']

0      Australia and New Zealand
1      Australia and New Zealand
2      Australia and New Zealand
3      Australia and New Zealand
4      Australia and New Zealand
                 ...            
847               Western Europe
848               Western Europe
849               Western Europe
850               Western Europe
851               Western Europe
Name: region, Length: 852, dtype: object

### Finding unique values in a column

The `unique` method can be used to get unique values for a particular column. Here we apply it to the `region` column.

In [8]:
happiness['region'].unique()

array(['Australia and New Zealand', 'Central and Eastern Europe',
       'Eastern Asia', 'Latin America and Caribbean',
       'Middle East and Northern Africa', 'North America',
       'Southeastern Asia', 'Southern Asia', 'Sub-Saharan Africa',
       'Western Europe'], dtype=object)

`happiness['region'].unique()` returns an array, which we can iterate over.

In [9]:
for region in happiness['region'].unique():
    print(region)

Australia and New Zealand
Central and Eastern Europe
Eastern Asia
Latin America and Caribbean
Middle East and Northern Africa
North America
Southeastern Asia
Southern Asia
Sub-Saharan Africa
Western Europe


### Filtering rows

Filtering is another popular technique in Python pandas to learn for data science. It’s useful when we want to subset the data based on some condition(s).

For example, let's grab the rows with `happiness_score` > 7 in the `region` 'Western Europe'. 

The variable msk below is a boolean array with a length equal to the number of rows in the DataFrame. The array element is True when the condition is satisfied, and False otherwise.

In [10]:
msk = (happiness['happiness_score'] > 7) & (happiness['region'] == 'Western Europe')
msk

0      False
1      False
2      False
3      False
4      False
       ...  
847    False
848    False
849     True
850     True
851     True
Length: 852, dtype: bool

We can use this boolean array in the squared brackets of the DataFrame `happiness` to do the filtering.

In [11]:
happiness[(happiness['region'] == 'Western Europe') & (happiness['happiness_score'] > 7)]

country          region  happiness_rank  happiness_score  year
726         Austria  Western Europe              13           7.2000  2015
727         Austria  Western Europe              12           7.1190  2016
728         Austria  Western Europe              13           7.0060  2017
729         Austria  Western Europe              12           7.1390  2018
730         Austria  Western Europe              10           7.2460  2019
731         Austria  Western Europe               8           7.2942  2020
744         Denmark  Western Europe               3           7.5270  2015
745         Denmark  Western Europe               1           7.5260  2016
746         Denmark  Western Europe               2           7.5220  2017
747         Denmark  Western Europe               3           7.5550  2018
748         Denmark  Western Europe               2           7.6000  2019
749         Denmark  Western Europe               1           7.6456  2020
750         Finland  Western Europe               6           7.4060  2015
751         Finland  Western Europe               5           7.4130  2016
752         Finland  Western Europe               5           7.4690  2017
753         Finland  Western Europe               1           7.6320  2018
754         Finland  Western Europe               1           7.7690  2019
755         Finland  Western Europe               0           7.8087  2020
767         Germany  Western Europe              16           7.0758  2020
774         Iceland  Western Europe               2           7.5610  2015
775         Iceland  Western Europe               3           7.5010  2016
776         Iceland  Western Europe               3           7.5040  2017
777         Iceland  Western Europe               4           7.4950  2018
778         Iceland  Western Europe               4           7.4940  2019
779         Iceland  Western Europe               3           7.5045  2020
784         Ireland  Western Europe              16           7.0210  2019
785         Ireland  Western Europe              15           7.0937  2020
796      Luxembourg  Western Europe              14           7.0900  2019
797      Luxembourg  Western Europe               9           7.2375  2020
804     Netherlands  Western Europe               7           7.3780  2015
805     Netherlands  Western Europe               7           7.3390  2016
806     Netherlands  Western Europe               6           7.3770  2017
807     Netherlands  Western Europe               6           7.4410  2018
808     Netherlands  Western Europe               5           7.4880  2019
809     Netherlands  Western Europe               5           7.4489  2020
816          Norway  Western Europe               4           7.5220  2015
817          Norway  Western Europe               4           7.4980  2016
818          Norway  Western Europe               1           7.5370  2017
819          Norway  Western Europe               2           7.5940  2018
820          Norway  Western Europe               3           7.5540  2019
821          Norway  Western Europe               4           7.4880  2020
834          Sweden  Western Europe               8           7.3640  2015
835          Sweden  Western Europe              10           7.2910  2016
836          Sweden  Western Europe               9           7.2840  2017
837          Sweden  Western Europe               9           7.3140  2018
838          Sweden  Western Europe               7           7.3430  2019
839          Sweden  Western Europe               6           7.3535  2020
840     Switzerland  Western Europe               1           7.5870  2015
841     Switzerland  Western Europe               2           7.5090  2016
842     Switzerland  Western Europe               4           7.4940  2017
843     Switzerland  Western Europe               5           7.4870  2018
844     Switzerland  Western Europe               6           7.4800  2019
845     Switzerland  Western Europe  

As you may have noticed, we can use logical operators such as & (AND), | (OR) in filtering. It is good practice to put the parentheses around the conditions to specify the execution. Otherwise, it might not work as expected.

### Groupby

Before learning pandas groupby, let’s look at some aggregating functions.

Below are the functions we will be using in the course.

#### min

Gets the minimum value of a column

In [12]:
happiness['year'].min()

2015

#### max

Gets the maximum value of a column

In [13]:
happiness['year'].max()

2020

#### mean

Gets the mean (average) value of a column

In [14]:
happiness['happiness_score'].mean()

5.468441078411263

Instead of applying over the whole dataset, these aggregating functions are more powerful on groups.

pandas `groupby` is a popular manipulation tool for data science. It can be used to group large amounts of data and compute operations based on groups. The idea should be familiar to those who have used SQL’s `SELECT` and `GROUP BY` statement.

For example, we can get the mean of the column `happiness_score`, grouped by column `region`.

In [15]:
happiness_score_by_region = happiness.groupby('region')['happiness_score'].mean()
happiness_score_by_region

region
Australia and New Zealand          7.289033
Central and Eastern Europe         5.467569
Eastern Asia                       5.702057
Latin America and Caribbean        6.006472
Middle East and Northern Africa    5.351085
North America                      7.159892
Southeastern Asia                  5.398765
Southern Asia                      4.467486
Sub-Saharan Africa                 4.236903
Western Europe                     6.767029
Name: happiness_score, dtype: float64

Using the `type()` function, we can see that `happiness_score_by_region` is a Pandas `Series`.

In [16]:
type(happiness_score_by_region)

pandas.core.series.Series

If we want to convert it to a pandas DataFrame (which is more convenient for Plotly graphs), we could use the `reset_index` method.

In [17]:
happiness_score_by_region_df = happiness.groupby('region')['happiness_score'].mean().reset_index()
happiness_score_by_region_df

region  happiness_score
0        Australia and New Zealand         7.289033
1       Central and Eastern Europe         5.467569
2                     Eastern Asia         5.702057
3      Latin America and Caribbean         6.006472
4  Middle East and Northern Africa         5.351085
5                    North America         7.159892
6                Southeastern Asia         5.398765
7                    Southern Asia         4.467486
8               Sub-Saharan Africa         4.236903
9                   Western Europe         6.767029

In [18]:
type(happiness_score_by_region_df)

pandas.core.frame.DataFrame